explorer/update.go

```diff
@@ -0,0 +1,250 @@
+	CreateTransaction() error
+	Commit() error
+
+	AddSiacoinElement(sce types.SiacoinElement) error
+	AddSiafundElement(sfe types.SiafundElement) error
+	AddFileContractElement(fce types.FileContractElement) error
+	RemoveElement(id types.ElementID) error
+	AddChainStats(index types.ChainIndex, stats ChainStats) error
+	AddUnspentSiacoinElement(address types.Address, id types.ElementID) error
+	AddUnspentSiafundElement(address types.Address, id types.ElementID) error
+	RemoveUnspentSiacoinElement(address types.Address, id types.ElementID) error
+	RemoveUnspentSiafundElement(address types.Address, id types.ElementID) error
+	AddTransaction(txn types.Transaction, addresses []types.Address, block types.ChainIndex) error
+}
+
+// An Explorer contains a database storing information about blocks, outputs,
+// contracts.
+type Explorer struct {
+	db       Store
+	mu       sync.Mutex
+	tipStats ChainStats
+	vc       consensus.ValidationContext
+}
+
+// ProcessChainApplyUpdate implements chain.Subscriber.
+func (e *Explorer) ProcessChainApplyUpdate(cau *chain.ApplyUpdate, mayCommit bool) error {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	if err := e.db.CreateTransaction(); err != nil {
+		return err
+	}
+
+	stats := ChainStats{
+		Block:               cau.Block,
+		ValidationContext:   cau.Context,
+		ActiveContractCost:  e.tipStats.ActiveContractCost,
+		ActiveContractCount: e.tipStats.ActiveContractCount,
+		ActiveContractSize:  e.tipStats.ActiveContractSize,
+		TotalContractCost:   e.tipStats.TotalContractCost,
+		TotalContractSize:   e.tipStats.TotalContractSize,
+		TotalRevisionVolume: e.tipStats.TotalRevisionVolume,
+	}
+
+	for _, txn := range cau.Block.Transactions {
+		// get a unique list of all addresses involved in transaction
+		addrMap := make(map[types.Address]struct{})
+		for _, elem := range txn.SiacoinInputs {
+			addrMap[elem.Parent.Address] = struct{}{}
+		}
+		for _, elem := range txn.SiacoinOutputs {
+			addrMap[elem.Address] = struct{}{}
+		}
+		for _, elem := range txn.SiafundInputs {
+			addrMap[elem.Parent.Address] = struct{}{}
+		}
+		for _, elem := range txn.SiafundOutputs {
+			addrMap[elem.Address] = struct{}{}
+		}
+		var addrs []types.Address
+		for addr := range addrMap {
+			addrs = append(addrs, addr)
+		}
+		if err := e.db.AddTransaction(txn, addrs, stats.Block.Header.Index()); err != nil {
+			return err
+		}
+	}
+
+	for _, elem := range cau.SpentSiacoins {
+		if err := e.db.RemoveElement(elem.ID); err != nil {
+			return err
+		}
+		if err := e.db.RemoveUnspentSiacoinElement(elem.Address, elem.ID); err != nil {
+			return err
+		}
+		stats.SpentSiacoinsCount++
+	}
+	for _, elem := range cau.SpentSiafunds {
+		if err := e.db.RemoveElement(elem.ID); err != nil {
+			return err
+		}
+		if err := e.db.RemoveUnspentSiafundElement(elem.Address, elem.ID); err != nil {
+			return err
+		}
+		stats.SpentSiafundsCount++
+	}
+	for _, elem := range cau.ResolvedFileContracts {
```

Comment:
we'll have to account for all the different types of resolution here; see the docstring of `FileContractResolution`

Reply:
I got everything except this.  Do you mean accounting for renewals?